validator/service/src/validator_api.rs

```diff
@@ -2,71 +2,263 @@
     }
 
-    fn get_archive_blocks_impl(&self, _request: BlocksRequest) -> Result<ArchiveBlocks, RpcStatus> {
-        todo!()
+    fn get_archive_blocks_impl(
+        &self,
+        request: BlocksRequest,
+        logger: &Logger,
+    ) -> Result<ArchiveBlocks, RpcStatus> {
+        log::trace!(
+            logger,
+            "get_archive_blocks(offset:{}, limit:{})",
+            request.offset,
+            request.limit
+        );
+
+        if request.limit > MAX_BLOCKS_PER_REQUEST {
+            return Err(rpc_invalid_arg_error("get_archive_blocks", "limit", logger));
+        }
+
+        let start_index = request.offset;
+        let end_index = request.offset + request.limit as u64;
+
+        // Get "persistence type" blocks.
+        let mut blocks_data = Vec::new();
+        for block_index in start_index..end_index {
+            match self.ledger_db.get_block_data(block_index) {
+                Ok(block_data) => blocks_data.push(block_data),
+                Err(mc_ledger_db::Error::NotFound) => {
+                    // This is okay - it means we have reached the last block in the ledger in the
+                    // previous loop iteration.
+                    break;
+                }
+                Err(error) => {
+                    log::error!(logger, "Error getting block {}: {:?}", block_index, error);
+                    return Err(rpc_database_err(error, logger));
+                }
+            }
+        }
+
+        Ok(ArchiveBlocks::from(&blocks_data[..]))
     }
 
-    fn propose_tx_impl(&self, _request: Tx) -> Result<ProposeTxResponse, RpcStatus> {
-        todo!()
+    fn propose_tx_impl(&self, tx: Tx, logger: &Logger) -> Result<ProposeTxResponse, RpcStatus> {
+        // Convert Protobuf/GRPC Tx to Prost Tx
+        let tx = mc_transaction_core::tx::Tx::try_from(&tx)
+            .map_err(|_| rpc_invalid_arg_error("propose_tx", "tx", logger))?;
+
+        // Figure out which node to submit to
+        let responder_ids = self.conn_manager.responder_ids();
+        if responder_ids.is_empty() {
+            return Err(rpc_internal_error("propose_tx", "no peers", logger));
+        }
+
+        let idx = self.submit_node_offset.fetch_add(1, Ordering::SeqCst);
+        let responder_id = &responder_ids[idx % responder_ids.len()];
+
+        // Submit.
+        let tx_propose_result = self
+            .conn_manager
+            .conn(responder_id)
+            .ok_or_else(|| rpc_internal_error("propose_tx", "conn not found", logger))?
+            .propose_tx(&tx, std::iter::empty());
+
+        // Convert to GRPC response.
+        let mut result = ProposeTxResponse::new();
+
+        match tx_propose_result {
+            Ok(block_count) => {
+                result.set_block_count(block_count);
+                Ok(())
+            }
+
+            Err(RetryError::Operation { error, .. }) => {
+                match error {
+                    ConnectionError::TransactionValidation(err) => {
```

Comment:
Is there a specific reason that the Transaction validation error has a different format and behaviour than the other 2?

Reply:
Yes. This is because I am trying to preserve the semantics of the ClientTxPropose call. This API call can fail either by returning a `TransactionValidation` error which is a valid GRPC response encapsulated in https://github.com/mobilecoinfoundation/mobilecoin/blob/bdd5ded7aff9b8a86bd10c568a1f2bcf1ee20d27/consensus/api/proto/consensus_common.proto#L80, or the GRPC call itself can fail for various reasons (attestation issues, network issues, etc).
The plan is to try and implement a "ThinClient" like you started doing, and I want to be able to return proper error results from it as much as possible. This special treatment of `TransactionValidation`, as well as the `Cipher` and `Attestation` variants will hopefully do that. Relaying the correct error value and not coercing everything into a generic "internal error" or something of that nature is important for correct client behavior.